lib/Dialect/HW/Transforms/HWVectorization.cpp

```diff
@@ -0,0 +1,879 @@
+//===- HWVectorization.cpp - HW Vectorization Pass ----------------*- C++
+//-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass performs structural vectorization of hardware modules,
+// merging scalar bit-level assignments into vectorized operations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombDialect.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWDialect.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "circt/Dialect/SV/SVDialect.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+#include <algorithm>
+
+namespace circt {
+namespace hw {
+#define GEN_PASS_DEF_HWVECTORIZATION
+#include "circt/Dialect/HW/Passes.h.inc"
+} // namespace hw
+} // namespace circt
+
+const int INDEX_SENTINEL_VALUE = -1;
+const int TOMBSTONE_SENTINEL_VALUE = -1;
+
+struct bit {
+  mlir::Value source;
+  int index;
+
+  bit(mlir::Value source, int index);
+  bit();
+  bit(const bit &other);
+
+  bit &operator=(const bit &other);
+  bool operator==(const bit &other) const;
+
+  bool left_adjacent(const bit &other);
```

Comment:
nit: please use camelBack (as noted in MLIR style guide https://mlir.llvm.org/getting_started/DeveloperGuide/#style-guide)